File: 0948-bag-of-tokens/0948-bag-of-tokens.js
/**
 * @param {number[]} tokens
 * @param {number} power
 * @return {number}
 */
var bagOfTokensScore = function(tokens, power) {
    tokens.sort((a, b) => a - b)
    let maxScore = 0, curScore = 0, l = 0, r = tokens.length-1
    while (l <= r) {
        if (power >= tokens[l]) {
            curScore++
            power -= tokens[l]
            l++
            maxScore = Math.max(maxScore, curScore)
        } else if (power < tokens[l] && curScore > 0) {
            curScore--
            power += tokens[r]
            r--
        } else {
            return maxScore
        }
    }
    return maxScore
};

File: 0948-bag-of-tokens/0948-bag-of-tokens.py
class Solution:
    def bagOfTokensScore(self, tokens: List[int], power: int) -> int:
        tokens.sort()
        res, score, l, r = 0, 0, 0, len(tokens)-1
        while l <= r:
            if power >= tokens[l]:
                power -= tokens[l]
                l += 1
                score += 1
                res = max(res, score)
            elif power < tokens[l] and score > 0:
                power += tokens[r]
                r -= 1
                score -= 1
            else:
                return res
        return res